client/X11/xf_cliprdr.c

```diff
@@ -1583,6 +1598,247 @@ static char* xf_cliprdr_fuse_split_basename(char* name, int len)
+
+		char* baseName = NULL;
+		UINT64 ticks;
+		xfCliprdrFuseInode* parent;
+
+		dirName = NULL;
+
+		inode->lindex = lindex;
+		inode->ino = lindex + 2;
+
+		if (split_point == NULL)
+		{
+			baseName = _strdup(curName);
+			if (!baseName)
+			{
+				goto error6;
+			}
+			inode->parent_ino = 1;
+			inode->name = baseName;
+			if (ArrayList_Add(rootNode->child_inos, (void*)inode->ino) < 0)
+			{
+				goto error6;
+			}
+		}
+		else
+		{
+			dirName = calloc(split_point - curName + 1, sizeof(char));
+			if (!dirName)
+			{
+				goto error6;
+			}
+			_snprintf(dirName, split_point - curName + 1, "%s", curName);
+			/* drop last '\\' */
+			baseName = _strdup(split_point + 1);
+			if (!baseName)
+			{
+				goto error7;
+			}
+
+			parent = (xfCliprdrFuseInode*)HashTable_GetItemValue(mapDir, dirName);
+			if (!parent)
+			{
+				goto error7;
+			}
+			inode->parent_ino = parent->ino;
+			inode->name = baseName;
+			if (ArrayList_Add(parent->child_inos, (void*)inode->ino) < 0)
+			{
+				goto error7;
+			}
+			free(dirName);
+		}
+		/* TODO: check FD_ATTRIBUTES in dwFlags
+		    However if this flag is not valid how can we determine file/folder?
+		*/
+		if ((descriptor->dwFileAttributes & FILE_ATTRIBUTE_DIRECTORY) != 0)
+		{
+			inode->st_mode = S_IFDIR | 0755;
+			inode->child_inos = ArrayList_New(TRUE);
+			if (!inode->child_inos)
+			{
+				goto error6;
+			}
+			inode->st_size = 0;
+			inode->size_set = TRUE;
+			char* tmpName = _strdup(curName);
+			if (!tmpName)
+			{
+				goto error6;
+			}
+			if (HashTable_Add(mapDir, tmpName, inode) < 0)
+			{
+				free(tmpName);
+				goto error6;
+			}
+		}
+		else
+		{
+			inode->st_mode = S_IFREG | 0644;
+			if ((descriptor->dwFlags & FD_FILESIZE) != 0)
+			{
+				inode->st_size = (((UINT64)descriptor->nFileSizeHigh) << 32) |
+				                 ((UINT64)descriptor->nFileSizeLow);
+				inode->size_set = TRUE;
+			}
+			else
+			{
+				inode->size_set = FALSE;
+			}
+		}
+
+		free(curName);
+
+		if ((descriptor->dwFlags & FD_WRITESTIME) != 0)
+		{
+			ticks = (((UINT64)descriptor->ftLastWriteTime.dwHighDateTime << 32) |
+			         ((UINT64)descriptor->ftLastWriteTime.dwLowDateTime)) -
+			        WIN32_FILETIME_TO_UNIX_EPOCH_USEC;
+			inode->st_mtim.tv_sec = ticks / 10000000ULL;
+			/* tv_nsec Not used for now */
+			inode->st_mtim.tv_nsec = ticks % 10000000ULL;
+		}
+		else
+		{
+			inode->st_mtim.tv_sec = time(NULL);
+			inode->st_mtim.tv_nsec = 0;
+		}
+
+		if (ArrayList_Add(clipboard->ino_list, inode) < 0)
+		{
+			goto error5;
+		}
+	}
+	ArrayList_Unlock(clipboard->ino_list);
+	free(descriptor);
+	HashTable_Free(mapDir);
+
+	Stream_Free(s, FALSE);
+	return TRUE;
+error7:
+	free(dirName);
+error6:
+	free(curName);
+error5:
+	xf_cliprdr_fuse_inode_free(inode);
+error4:
+	free(descriptor);
+error3:
+	HashTable_Free(mapDir);
+error2:
+	ArrayList_Clear(clipboard->ino_list);
+error:
+	ArrayList_Unlock(clipboard->ino_list);
+	Stream_Free(s, FALSE);
+	return FALSE;
```

Comment:
This is a sign that there are too many things going on this function. It's very easy to mess up something here. Even if it's correct now, someone will edit it later and they could break something down: jump to an incorrect error label, forget to free or unlock something, etc.

I believe this could be improved by splitting this function into several parts where each is responsible for one or maybe a couple of resources and associated error handling. A function that allocates and frees `wStream`. A function that keeps a track of `ino_list`, its lock, and maybe the `HashTable`. A function that reads through the `FILEDESCRIPTORW` list and deals with filenames. A function that finally inserts `xfCliprdrFuseInode` objects. Or something like that.